packages/client/src/helpers.ts

```diff
@@ -0,0 +1,55 @@
+import { CommonPaginatedResultInfoFragment } from './graphql/fragments.generated';
+
+type PaginatedQueryData<K = unknown[]> = {
+  data: {
+    result: {
+      pageInfo: CommonPaginatedResultInfoFragment;
+      items: K;
+    };
+  };
+};
+
+type InferPaginatedItemsType<T extends PaginatedQueryData<unknown>> = T extends PaginatedQueryData<
+  infer R
+>
+  ? R
+  : never;
+
+export type PaginatedResult<T> = T & {
+  next(): Promise<PaginatedResult<T> | null>;
+  prev(): Promise<PaginatedResult<T> | null>;
+};
+
+export async function buildPaginatedQueryResult<
+  V,
+  T extends PaginatedQueryData<K>,
+  K = InferPaginatedItemsType<T>,
+>(queryFn: (variables: V) => Promise<K>, variables: V): Promise<PaginatedResult<K>> {
+  const result = await queryFn(variables);
+  const paginatedResult = result as PaginatedQueryData;
```

Comment:
@krzysu I couldn't figure out why `result` wasn't being typed as being extending `PaginatedQueryData` to use in the `next` and `prev` functions. Let me know if you have any ideas!

Reply:
Would refactor to something like below so types are nicely maintained
```

type PaginatedQueryData<Item> = {
  pageInfo: CommonPaginatedResultInfoFragment;
  items: Item[];
};

export type PaginatedResult<T> = PaginatedQueryData<T> & {
  next(): Promise<PaginatedResult<T> | null>;
  prev(): Promise<PaginatedResult<T> | null>;
};

export async function buildPaginatedQueryResult<V, R>(
  queryFn: (variables: V) => Promise<PaginatedQueryData<R>>,
  variables: V,
): Promise<PaginatedResult<R>> {
  const result = await queryFn(variables);

  return {
    ...result,
    async next() {
      if (result.pageInfo.next) {
        return buildPaginatedQueryResult(queryFn, {
          ...variables,
          cursor: result.pageInfo.next,
        });
      }

      return null;
    },

    async prev() {
      if (result.pageInfo.prev) {
        return buildPaginatedQueryResult(queryFn, {
          ...variables,
          cursor: result.pageInfo.prev,
        });
      }

      return null;
    },
  };
}
```